cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/Language.kt

```diff
@@ -273,6 +356,60 @@ abstract class Language<T : LanguageFrontend<*, *>> : Node() {
             signature.size == targetSignature.size
         }
     }
+
+    /**
+     * This functions gives the language a chance to refine the results of a
+     * [ScopeManager.resolveCall]. The default implementation will follow the following heuristic:
+     * - If the list of viableFunctions is empty, we can directly return.
+     * - If we have only one item in viableFunctions, we can take it.
+     * - Next, we can check for direct matches, meaning that they have a SignatureResult that only
+     *   has DirectMatchNoCastNeeded casts.
+     * - Lastly, if we have not direct matches, we need to sort the viable functions using a simple
+     *   ranking. The function(s) will the best (lowest) ranking will be chosen as the best. The
+     *   ranking is determined by the "depth" of all cast results in the signature results.
+     */
+    open fun bestViableResolution(
+        result: ScopeManager.CallResolutionResult
+    ): List<FunctionDeclaration> {
+        // No need to do anything, if the list of viable functions is empty
+        if (result.viableFunctions.isEmpty()) {
+            return listOf()
+        }
+
+        // If we already have only one viable result, we can just go back
+        val single = result.viableFunctions.singleOrNull()
+        if (single != null) {
+            return listOf(single)
+        }
+
+        // Check for direct matches. Let's hope there is only one, otherwise we have an ambiguous
+        // result
+        val directMatches = result.signatureResults.entries.filter { it.value.isDirectMatch }
+        return if (directMatches.size > 1) {
+            // This is an ambiguous result
+            result.type = ScopeManager.CallResolutionResult.Type.AMBIGUOUS
+            // Let's return all direct matches
+            directMatches.map { it.key }
+        } else if (directMatches.size == 1) {
+            // Let's return the single direct match
+            listOf(directMatches.first().key)
```

Comment:
I'd somehow prefer to set the `result.type` here instead of having the default `OK` set in the `ScopeManager`

Reply:
I can try, the only issue is that we need to set the PROBLEMATIC already in resolveCall when we have a language that does not allow function overloading but more than one candidate function was returned. We are basically not allowed to "recover" from the problematic state, so that's why I handle it in resolveCall.

Open to suggestions :)

Reply:
Hmm but couldn't we currently recover from it if there were multiple direct matches (which would be very weird) or if there is no direct match but several equally-ranked matches? Then, we should probably stop the matching after l. 796 in ScopeManager and set the remaining fields there or at least, we should not enter this function (which we do at the moment)?

Reply:
> Hmm but couldn't we currently recover from it if there were multiple direct matches (which would be very weird) or if there is no direct match but several equally-ranked matches? 

Ah you are correct, in this case we would "upgrade" from problematic to ambiguous, which is not intended.

> Then, we should probably stop the matching after l. 796 in ScopeManager and set the remaining fields there or at least, we should not enter this function (which we do at the moment)?

I think it's probably easiest if we do not enter bestViableResolution in this case then. That's how I implemented it now.